src/nlr/LPFormulator.cpp

```diff
@@ -61,93 +64,259 @@ double LPFormulator::solveLPRelaxation( const Map<unsigned, Layer *> &layers,
+                throw InfeasibleQueryException();
+
+            if ( !gurobi.optimal() )
+                throw NLRError( NLRError::UNEXPECTED_RETURN_STATUS_FROM_GUROBI );
+
+            gurobi.extractSolution( dontCare, ub );
+
+            // If the bound is tighter, store it
+            if ( ub < currentUb )
+            {
+                gurobi.setUpperBound( variableName, ub );
+
+                if ( FloatUtils::isPositive( currentUb ) &&
+                     !FloatUtils::isPositive( ub ) )
+                    ++signChanges;
+
+                layer->setUb( j, ub );
+                _layerOwner->receiveTighterBound( Tightening( variable,
+                                                              ub,
+                                                              Tightening::UB ) );
+                ++tighterBoundCounter;
+
+                if ( _cutoffInUse && ub < _cutoffValue )
+                {
+                    ++cutoffs;
+                    continue;
+                }
+            }
+
+            // Minimize
+            gurobi.reset();
+            gurobi.setCost( terms );
+            gurobi.solve();
+            if ( gurobi.infeasbile() )
+                throw InfeasibleQueryException();
 
-            // Store the new bounds if they are tighter
-            if ( lb > layer.second->getLb( i ) )
+            if ( !gurobi.optimal() )
+                throw NLRError( NLRError::UNEXPECTED_RETURN_STATUS_FROM_GUROBI );
+
+            gurobi.extractSolution( dontCare, lb );
+
+            // If the bound is tighter, store it
+            if ( lb > currentLb )
             {
-                if ( FloatUtils::isNegative( layer.second->getLb( i ) ) &&
+                gurobi.setLowerBound( variableName, lb );
+
+                if ( FloatUtils::isNegative( currentLb ) &&
                      !FloatUtils::isNegative( lb ) )
                     ++signChanges;
 
-                layer.second->setLb( i, lb );
+                layer->setLb( j, lb );
                 _layerOwner->receiveTighterBound( Tightening( variable,
                                                               lb,
                                                               Tightening::LB ) );
                 ++tighterBoundCounter;
+
+                if ( _cutoffInUse && lb > _cutoffValue )
+                {
+                    ++cutoffs;
+                    continue;
+                }
             }
+        }
+    }
 
-            if ( ub < layer.second->getUb( i ) )
+    gurobiEnd = TimeUtils::sampleMicro();
+
+    LPFormulator_LOG( Stringf( "Number of tighter bounds found by Gurobi: %u. Sign changes: %u. Cutoffs: %u\n",
+                               tighterBoundCounter, signChanges, cutoffs ).ascii() );
+    LPFormulator_LOG( Stringf( "Seconds spent Gurobiing: %llu\n", TimeUtils::timePassed( gurobiStart, gurobiEnd ) / 1000000 ).ascii() );
+}
+
+void LPFormulator::optimizeBoundsWithLpRelaxation( const Map<unsigned, Layer *> &layers )
```

Comment:
Is this method called anywhere?

Reply:
Added a configuration option that controls with of the 4 methods gets called (LP/MILP x Inc/non-Inc)